Test PropStorageInfo lookup

scober opened the pull request:
@ddn0 correctly pointed that I should have added some tests for `find_node_prop_info()` and friends. And I especially should have added some tests after I found a bug in the first implementation. While writing those tests I realized that `RDGCore` wasn't a particularly good home for those functions anyway.

So.

Move `find_*_prop_info()` from `RDGCore` to `RDGPartHeader` so that they are now in the same class as the data structures they operate on. And add a test for those functions as well as some functions that modify the `PropStorageInfo` lists.

Commit:
Cleanup RDGPartHeader interface

Change set_part_properties() to set_part_prop_info_list() to match
set_node_prop_info_list() and set_edge_prop_info_list().

Commit:
Move search functions closer to data structures

find_node_prop_info(), find_edge_prop_info(), and find_part_prop_info()
all search over data structures in RDGPartHeader. So move their
implementations out of RDGCore and into RDGPartHeader. This also allows
RDGPartHeader to re-use those search functions internally.

Commit:
Add RDGPartHeader test

Add a unit test for RDGPartHeader so it can be tested independently from
the other components of the system.

e-mcginnis reviewed (approved):
LGTM
Thanks for adding this test Simon!

witchel commented on libtsuba/test/rdg-part-header.cpp:
```diff
@@ -0,0 +1,97 @@
+  tsuba::PropStorageInfo updated_edge_prop("value", arrow::large_binary());
+  under_test.UpsertEdgePropStorageInfo(std::move(updated_edge_prop));
+
+  KATANA_LOG_ASSERT(under_test.node_prop_info_list().empty());
```
Shouldn't this be `!`?

scober replied:
This is tightly coupled to rmat15 as an input. I will add some comments/asserts indicating that. But with that in mind, rmat15 has one edge property and no node properties. So this is the correct assert.

witchel replied:
Cool.  Comment is worthwhile.

witchel commented on libtsuba/test/rdg-part-header.cpp:
```diff
@@ -0,0 +1,97 @@
+      under_test.find_edge_prop_info("value")->IsAbsent() == true);
+  KATANA_LOG_ASSERT(under_test.find_edge_prop_info("not value") != nullptr);
+  KATANA_LOG_ASSERT(
+      under_test.find_edge_prop_info("not value")->IsDirty() == true);
```
Why is "value" absent, while "not value" is dirty?

scober replied:
Nice catch! Turns out this was a bug, I am pushing a fix now.

witchel commented on libtsuba/test/rdg-part-header.cpp:
```diff
@@ -0,0 +1,97 @@
+
+  // remove everything
+  KATANA_CHECKED(under_test.RemoveEdgeProperty("not value"));
+  under_test.RemoveEdgeProperty(0);
```
I'd think you'd have to remove both of them.

scober replied:
I tried to remove one by name and one by index.

witchel reviewed (approved):
Nice tests.

I wonder about some of the asserts, but if they pass, they pass.

Commit:
PropStorageInfo: fix state machine transition bug

Is*() functions report state, Was*() functions indicate that state
should transition. Fix the Upsert*() family to use WasModified() (to
indicate that the property should be marked dirty) instead of IsDirty()
(which reports the current state of the property). Also ensure that the
property is marked dirty in both cases (insert and update).
